torch/distributed/_spmd/graph_optimization.py

```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+
+    def populate_outputs(self):
+        # Populate the existing output lists from the graph.
+        self.outputs = [
+            self.add_node for _ in cast(Tuple[Any, ...], self.add_node.args[0])
+        ]
+        for updated_arg in self.add_node.users:
+            assert (
+                updated_arg.target == operator.getitem
+            ), f"Unexpected node target {updated_arg.target}"
+            idx = cast(int, updated_arg.args[1])
+            output_copy = next(iter(updated_arg.users))
+            assert str(output_copy.target).startswith(
+                "aten.copy_"
+            ), f"The execpted output node is different, {str(output_copy.target)}"
+            self.outputs[idx] = output_copy
+        for i, output in enumerate(self.outputs):
+            assert output != self.add_node, f"{i}th output is not replaced."
+
+    def __post_init__(self):
+        if self.outputs:
+            return
+
+        if self.generate_output:
+            self.generate_outputs()
+        else:
+            self.populate_outputs()
+
+
+@dataclass(unsafe_hash=True)
+class FusedOptimizerBlock:
+    step: ForeachAddBlock
+    optim: FusedAdamBlock
+
+
+def get_fused_optimizer_block(optim_node: fx.Node) -> FusedOptimizerBlock:
+    """
+    Given a fused optimizer node and return the FusedOptimizerBlock.
+    """
+    MAX_STEP_DISTANCE = 5
+    # Find the step (foreach_add)
+    nodes = collections.deque([optim_node, None])
+    step_node = optim_node
+    distance = 0
+    while nodes and distance < MAX_STEP_DISTANCE:
+        node = nodes.popleft()
+        if node is None:
+            distance += 1
+            if nodes:
+                nodes.append(None)
+            continue
+        if node.op == OP.CALL_FUNCTION and str(node.target).startswith(
+            "aten._foreach_add"
+        ):
+            step_node = node
+            break
+        else:
+            flatten_args, _ = tree_flatten((node.args, node.kwargs))
+            for ancestor in flatten_args:
+                if isinstance(ancestor, fx.Node):
+                    nodes.append(ancestor)
+    if step_node == optim_node:
+        raise RuntimeError(
+            "Cannot find step node (foreach_add) for the optimizer node "
+            f"{optim_node}. The API design does not match the tracing graph."
+        )
+
+    step = ForeachAddBlock(step_node, generate_output=False)
+    optim = FusedAdamBlock(optim_node, generate_output=False)
+    return FusedOptimizerBlock(step, optim)
+
+
+def get_all_fused_optimizer_blocks(
+    gm: IterGraphModule, optim_ops: Union[Tuple[str, ...], str]
+) -> List[FusedOptimizerBlock]:
+    """
+    Find all the FusedOptimizerBlock that the optimizer operators are in
+    `optim_ops`.
+    """
+    return [
+        get_fused_optimizer_block(node)
+        for node in gm.graph.nodes
+        if node.name.startswith(optim_ops)
+    ]
+
+
+def _split_fused_adam(
```

Comment:
cc @yifuwang , IIUC, this needs to happen before we re-inplace fused Adam

Reply:
Yes